tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/variable/MutableVariable.java

```diff
@@ -0,0 +1,246 @@
+ limitations under the License.
+ ==============================================================================
+ */
+package org.tensorflow.variable;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.function.Supplier;
+import org.tensorflow.Operand;
+import org.tensorflow.ndarray.Shape;
+import org.tensorflow.op.Op;
+import org.tensorflow.op.Operands;
+import org.tensorflow.op.Scope;
+import org.tensorflow.op.core.AssignAddVariableOp;
+import org.tensorflow.op.core.AssignSubVariableOp;
+import org.tensorflow.op.core.AssignVariableOp;
+import org.tensorflow.op.core.Init;
+import org.tensorflow.op.core.IsVariableInitialized;
+import org.tensorflow.op.core.ReadVariableOp;
+import org.tensorflow.op.core.VarHandleOp;
+import org.tensorflow.op.core.VarHandleOp.Options;
+import org.tensorflow.proto.framework.DataType;
+import org.tensorflow.types.TBool;
+import org.tensorflow.types.family.TType;
+
+/**
+ * The implementation of {@link Variable}, with mutation methods.
+ *
+ * @see Variable
+ */
+public class MutableVariable<T extends TType> implements Variable<T> {
+
+  private final Scope initialScope;
+  private final String name;
+
+  private final Shape shape;
+  private final DataType dataType;
+  private final Class<T> tType;
+  private final VarHandleOp handle;
+
+  private IsVariableInitialized isInitializedOp = null;
+  private Op initializationOp = null;
+  private ReadVariableOp<T> cachedRead = null;
+  private Op lastAssign = null;
+
+  private boolean hasInitialized = false;
+
+  protected MutableVariable(Scope scope, Shape shape, Class<T> dataType) {
+    this.shape = shape;
+    this.dataType = Operands.toDataType(dataType);
+    this.tType = dataType;
+
+    this.name = scope.makeOpName("Variable");
+
+    scope = scope.withName(null);
+    this.initialScope = scope.withSubScope(this.name);
+
+    VarHandleOp.Options[] options;
+
+    if (scope.env().isGraph()) {
+      options = new Options[]{VarHandleOp.sharedName(this.name)};
+    } else {
+      options = new Options[0];
+    }
+
+    this.handle = VarHandleOp.create(initialScope.withName(name), dataType, shape, options);
+
+    scope.env().registerVariable(this);
+  }
+
+  @Override
+  public Shape getShape() {
+    return shape;
+  }
+
+  @Override
+  public DataType getDataType() {
+    return dataType;
+  }
+
+  @Override
+  public boolean isInitialized() {
+    return hasInitialized;
+  }
+
+  @Override
+  public String getName() {
+    return name;
+  }
+
+  @Override
+  public Operand<T> value(Scope scope) {
+    if (!hasInitialized) {
+      throw new IllegalStateException("Variable has not been initialized, can not get.");
+    }
+    if (cachedRead == null) {
+      if (lastAssign != null) {
+        scope = scope.withControlDependencies(Collections.singletonList(lastAssign));
+      }
+      cachedRead = ReadVariableOp.create(scope, handle, tType);
+    }
+    return cachedRead;
```

Comment:
This method has data races. In Python they don't need to worry about it, but we should at least consider what the possible behaviours are in Java.

Reply:
Yeah, I should use a local variable there.  Do you think using `AtomicReference` as well is worth it?  I'm not sure how much of the rest is thread safe.